### DISCUSSION TITLE
Enhancement: Attribute-based variants

### DISCUSSION THREAD

**Discussion by jshimkoski** [tailwindlabs/tailwindcss#7564] [Ideas]
Imagine you are using something like [vue-lazyload](https://github.com/hilongjw/vue-lazyload), which adds a `lazy` attribute to an element during certain state changes, for example:

```
<img
  v-lazy="..."
  lazy="loading" // this is auto generated for us
  class="transition-opacity duration-500 ease-in-out"
/>
```

At the current time, you need to write custom CSS to handle this:

```
<style lang="postcss" scoped>
img[lazy=loading] {
  @apply opacity-0;
}

img[lazy=loaded] {
  @apply opacity-100;
}
</style>
```

It would be great if Tailwind JIT could handle these scenarios, kinda like:

```
<img
  src="..."
  lazy="loaded"
  class="transition-opacity duration-500 ease-in-out [lazy=loading]:opacity-0 [lazy=loaded]:opacity-100"
/>
```

Obviously, the API would need to be finessed for readability and to allow for values with spaces and such.

But, honestly, this would be killer and would eliminate most of the remaining moments that we need to write custom CSS in our projects. The only other exception I can think of, off the top of my head, would be dynamically injected child content, and I'm not so sure how that could be handled cleanly.

**Comment by jshimkoski** [comment]
Something else, which, admittedly, is kinda bonkers, but would have its use cases:

```
class="{#el_id}:mt-0 {.class_name}:mt-0 {[attr_name=attr_value]}:mt-0 {#parent > p.dynamic_child}:mt-0"
```

That would generate:

```
#el_id {
  margin-top: 0;
}

.class_name {
  margin-top: 0;
}

[attr_name=attr_value] {
  margin-top: 0;
}

#parent > p.dynamic_child {
  margin-top: 0;
}
```

Possible use cases:

* When element is added to the DOM it changes the CSS of another element 
* Enforces a higher CSS specificity.
* Allows for the handling of the dynamic attributes as mentioned originally
* Could specify dynamically generated children

Again, this might be overkill but a neat idea nonetheless.

**Comment by adamwathan** [comment]
This is all pretty evil shit which means I'm excited to play with it and see if I can get it working 😅 I'll try to find time to tinker with this and let you know what progress I make! It's well aligned with this idea I had:

```html
<div class="md:hover:{mix-blend-mode:multiply}">
```

**Comment by jshimkoski** [comment]
This could be used to style before and after pseudo elements as well.

```
class="test {.test::before}:inline-block {.test::before}:{content:empty} {.test::before}:mx-[2px]"
```

Turns into:

```
.test::before {
  display: inline-block;
  content: '';
  margin-left: 2px;
  margin-right: 2px;
}
```

**Comment by RobinMalfait** [comment]
Hey! Doing a big GitHub clean-up this week so going to move this over to a discussion. Still excited to play with this idea though 🥳
